src/environmentd/tests/server.rs

```diff
@@ -4488,3 +4489,172 @@ async fn test_cert_reloading() {
+        let new_mv_id: String = client
+            .query_one(
+                &format!("SELECT id FROM mz_materialized_views WHERE name = '{mv_name}'"),
+                &[],
+            )
+            .await
+            .expect("failed to query")
+            .get(0);
+        // Assert that the materialized view did not get a new ID.
+        assert_eq!(new_mv_id, mv_id);
+
+        // TODO(jkosh44) This does not work, the old server halting will terminate the entire
+        // process. We could fork the process and run the read-only server in the forked process,
+        // but that's probably not worth it. A better fix is to re-write this as a platform check.
+        Retry::default()
+            .max_duration(Duration::from_secs(30))
+            .retry_async(|_state| async {
+                let response = server.promote().await;
+                if response.status().is_success() {
+                    Ok(())
+                } else {
+                    Err(response)
+                }
+            })
+            .await
+            .unwrap();
+    }
+
+    // We need to promote and restart the server to see the new shard IDs, because the read-only
+    // instance cannot write to mz_internal.mz_storage_shards to update the shard mapping.
+    {
+        let server = harness
+            .clone()
+            .with_deploy_generation(second_generation)
+            .start()
+            .await;
+        let client = server.connect().await.unwrap();
+
+        let new_mz_tables_shard: String = Retry::default()
+            .max_duration(Duration::from_secs(30))
+            .retry_async(|_state| async {
+                let new_mz_tables_shard: Result<String, _> = client
+                    .query_one(&format!("SELECT shard_id FROM mz_internal.mz_storage_shards WHERE object_id = '{mz_tables_id}'"), &[])
+                    .await
+                    .map(|row| row.get(0));
+                match new_mz_tables_shard {
+                    Ok(new_mz_tables_shard) if new_mz_tables_shard == mz_tables_shard => Err("mz_tables shard was not updated".to_string()),
+                    Ok(new_mz_tables_shard) => Ok(new_mz_tables_shard),
+                    Err(e) => Err(e.to_string()),
+                }
+            })
+            .await
+            .unwrap();
+        // Assert that the table did get a new shard.
+        assert_ne!(new_mz_tables_shard, mz_tables_shard);
+        let new_mv_shard: String = Retry::default()
+            .max_duration(Duration::from_secs(30))
+            .retry_async(|_state| async {
+                client
+                    .query_one(&format!("SELECT shard_id FROM mz_internal.mz_storage_shards WHERE object_id = '{mv_id}'"), &[])
+                    .await
+                    .map(|row| row.get(0))
+            })
+            .await
+            .unwrap();
+        // Assert that the materialized view did not get a new shard.
+        assert_eq!(new_mv_shard, mv_shard);
+    }
+}
```

Comment:
This test fails because halting the servers calls `libc::_exit` which exits the entire process and fails the test. Running through these steps manually behaves as expected in the test. I'm working on coming up with a new test for this.

Reply:
Is it worth making the halt operation configurable? For example could we configure it to return from the Coordinator serve method in tests instead of exiting?